### PR TITLE
DEV: Introduce ember-decorators addon

### PR DESCRIPTION
**Pull request by davidtaylorhq**
**DEV: Introduce ember-decorators addon**

This provides Classic Component decorators for use with native class syntax (e.g. `@tagName`), and also provides native-class-compatible decorators for `@on` and `@observes`.

---

**DEV: Add native class shims for `on`/`observes` decorators**

### CODE REVIEW

File: app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js [hunk 1]
```diff
@@ -175,4 +177,78 @@ module("Unit | Utils | decorators", function (hooks) {
 
     assert.strictEqual(stub.otherCounter, 1);
   });
+
+  test("@observes works via .extends and native class syntax", async function (assert) {
```
Comment (CvX) [inline app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js:181]: `.extends` -> `.extend` (that's the name of the Ember function, `extends` is the native class keyword 😅)

**Review by CvX — COMMENTED** [review]
That one ff failure is an unrelated flake? Haven't seen that one before

File: app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js (hunk 1 above)
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js]: Right 🤦 Fixed

Flake is unrelated to this PR - it's happening on `main` as well. Digging now 👀

File: app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js [hunk 2]
```diff
@@ -175,4 +177,78 @@ module("Unit | Utils | decorators", function (hooks) {
...
+
+    const extendsTest = ExtendsWithObserver.create();
+    extendsTest.set("value", "one");
+    await settled();
+    extendsTest.set("value", "two");
+    await settled();
+    assert.strictEqual(
+      extendsTest.counter,
+      2,
+      "observer triggered for extends"
```
Comment (CvX) [inline app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js:220]:
```suggestion
      "observer triggered for extend"
```

File: app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js [hunk 3]
```diff
@@ -175,4 +177,78 @@ module("Unit | Utils | decorators", function (hooks) {
...
+
+    const nativeClassTest = NativeClassWithOn.create();
+    assert.strictEqual(
+      nativeClassTest.counter,
+      1,
+      "on triggered for native class"
+    );
+
+    const extendsTest = ExtendsWithOn.create();
+    assert.strictEqual(extendsTest.counter, 1, "on triggered for extends");
```
Comment (CvX) [inline app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js:252]:
```suggestion
    assert.strictEqual(extendsTest.counter, 1, "on triggered for extend");
```

(vars and class names can stay as is 😉)

**Review by CvX — APPROVED** [review]

File: app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js (hunk 3 above)
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/tests/unit/utils/decorators-test.js]: Ok, I fixed all of them now (including class/var names). Thanks!

**Pull request merged**